contracts/evm/contracts/Oracle.sol

```diff
@@ -0,0 +1,258 @@
+            pairs[_pairs[i].id] = _pairs[i];
+        }
+    }
+
+    function setSourcesThreshold(uint256 threshold) external onlyOwner {
+        sourcesThreshold = threshold;
+    }
+
+    function updatePublisherRegistryAddress(
+        IPublisherRegistry newPublisherRegistryAddress
+    ) external onlyOwner {
+        publisherRegistry = newPublisherRegistryAddress;
+    }
+
+    function publishSpotEntry(SpotEntry calldata spotEntry) public {
+        _validateSenderForSource(spotEntry.base, msg.sender);
+        SpotEntry memory _latest = spotEntryStorage[spotEntry.pairId][
+            spotEntry.base.source
+        ];
+        _validateTimestamp(_latest, spotEntry);
+        spotEntryStorage[spotEntry.pairId][spotEntry.base.source] = spotEntry;
+
+        emit SubmittedSpotEntry(spotEntry);
+    }
+
+    function setCheckpoint(bytes32 pairId, AggregationMode aggregationMode)
+        public
+    {
+        bytes32[] memory sources = oracleSourcesStorage[pairId];
+        (
+            uint256 value,
+            ,
+            uint256 lastUpdatedTimestamp,
+            uint256 numSourcesAggregated
+        ) = getSpot(pairId, aggregationMode, sources);
+
+        require(
+            sourcesThreshold <= numSourcesAggregated,
+            "Does not meet threshold for aggreagated sources"
+        );
+
+        if (checkpointIndex[pairId] > 0) {
+            Checkpoint memory currentCheckpoint = checkpoints[pairId][
+                checkpointIndex[pairId] - 1
+            ];
+            require(
+                currentCheckpoint.timestamp < lastUpdatedTimestamp,
+                "stale"
+            );
+        }
+        Checkpoint memory newCheckpoint = Checkpoint(
+            lastUpdatedTimestamp,
+            value,
+            aggregationMode,
+            numSourcesAggregated
+        );
+
+        checkpointIndex[pairId]++;
+        checkpoints[pairId].push(newCheckpoint);
+
+        return ();
+    }
+
+    function publishSpotEntries(SpotEntry[] calldata spotEntries) public {
+        for (uint256 i = 0; i < spotEntries.length; i++) {
+            publishSpotEntry(spotEntries[i]);
+        }
+    }
+
+    function getSpot(
+        bytes32 pairId,
+        AggregationMode aggregationMode,
+        bytes32[] memory sources
+    )
+        public
+        view
+        returns (
+            uint256 price,
+            uint256 decimals,
+            uint256 lastUpdatedTimestamp,
+            uint256 numSourcesAggregated
+        )
+    {
+        (SpotEntry[] memory entries, uint256 length, ) = getSpotEntries(
+            pairId,
+            sources
+        );
+        if (entries.length == 0) {
+            return (0, 0, 0, 0);
+        }
+        uint256 price = _aggregateSpotEntries(entries);
+        uint256 decimals = _getSpotDecimals(pairId);
+        uint256 lastUpdatedTimestamp = _maxSpotTimestamp(entries);
+        return (price, decimals, lastUpdatedTimestamp, length);
```

Comment:
Maybe this is a Solidity thing I am unfamiliar with but should this just be `entries.length`?

Reply:
no because you need to know the size of the array when you create it, so the array has a max length and then some of the entries get filtered out.  so we need to track its length.

Reply:
so you need to say something like `uint[] myArr = new uint[](5);` and then I'm only using the first 3 entries.